### PR TITLE
Add gRPC communication between agent and collector

### PR DESCRIPTION
**Pull request by pavolloffay**
Resolves some parts of #773

Rebased and modified #903 and #1042 

* grpc server on the collector and reporter on the agent side
* sampling strategies over grpc
* grpc flags are marked as experimental. By default is used tchannel
* xdock uses grpc

### CODE REVIEW

File: model/proto/api_v2.proto [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    Batch batch = 1 [
+        (gogoproto.nullable) = false
+    ];
+}
+
+message PostSpansResponse {
+    bool ok = 1;
+}
+
+service CollectorService {
```
Comment (pavolloffay) [inline model/proto/api_v2.proto]: For now I have excluded query service. Let's do it in a separate PR.

File: cmd/agent/app/reporter/grpc/sampling_manager.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	}
+}
+
+// GetSamplingStrategy returns sampling strategies from collector.
+func (s *SamplingManager) GetSamplingStrategy(serviceName string) (*sampling.SamplingStrategyResponse, error) {
+	r, err := s.sampling.GetSamplingStrategy(context.Background(), &api_v2.SamplingStrategyParameters{ServiceName: serviceName})
+	if err != nil {
+		return nil, err
+	}
+	return jaeger.ConvertSamplingResponseFromDomain(r)
```
[the remark this answers is not in the document]
Comment (pavolloffay) [reply cmd/agent/app/reporter/grpc/sampling_manager.go]: definitely, tracked in https://github.com/jaegertracing/jaeger/issues/773#issuecomment-436602664

File: cmd/all-in-one/main.go [hunk 3]
```diff
@@ -216,38 +250,65 @@ func startCollector(
...
+		if err != nil {
+			logger.Fatal("Failed to listen on gRPC port", zap.Error(err))
+		}
+
+		grpclog.SetLoggerV2(grpclog.NewLoggerV2(ioutil.Discard, os.Stderr, os.Stderr))
+
+		grpcSrv := grpc.NewServer()
+		api_v2.RegisterCollectorServiceServer(grpcSrv, grpcHandler)
+		api_v2.RegisterSamplingManagerServer(grpcSrv, sampling.NewGRPCHandler(strategyStore))
+		logger.Info("Starting Jaeger Collector gRPC server", zap.Int("grpc-port", cOpts.CollectorGRPCPort))
```
[the remark this answers is not in the document]
Comment (pavolloffay) [reply cmd/all-in-one/main.go]: Fixed also in other mains it wasn't consistent on jaeger-

File: cmd/all-in-one/Dockerfile [hunk 4]
```diff
@@ -15,6 +15,9 @@ EXPOSE 5778
 # Collector HTTP
 EXPOSE 14268
 
+# Collector gRPC
+EXPOSE 14270
```
[the remark this answers is not in the document]
Comment (pavolloffay) [reply cmd/all-in-one/Dockerfile]: +1 it has to be changed. Do we have any preference?

File: cmd/all-in-one/main.go [hunk 5]
```diff
@@ -197,12 +210,33 @@ func startAgent(
 	}
 }
 
+func createCollectorProxy(
+	cOpts *collector.CollectorOptions,
+	opts *agentRep.Options,
+	tchanRep *agentTchanRep.Builder,
+	grpcRepOpts *agentGrpcRep.Options,
```
[the remark this answers is not in the document]
Comment (pavolloffay) [reply cmd/all-in-one/main.go]: Tchannel builder is a builder (with methods and create..) whereas others just holds configuration properties calling them builder seemed weird.

File: cmd/all-in-one/main.go [hunk 6]
```diff
@@ -216,38 +250,65 @@ func startCollector(
...
-	zipkinSpansHandler, jaegerBatchesHandler := spanBuilder.BuildHandlers()
-	server.Register(jc.NewTChanCollectorServer(jaegerBatchesHandler))
-	server.Register(zc.NewTChanZipkinCollectorServer(zipkinSpansHandler))
-	server.Register(sc.NewTChanSamplingManagerServer(samplingHandler))
-	portStr := ":" + strconv.Itoa(cOpts.CollectorPort)
-	listener, err := net.Listen("tcp", portStr)
-	if err != nil {
-		logger.Fatal("Unable to start listening on channel", zap.Error(err))
+
+	{
```
[the remark this answers is not in the document]
Comment (pavolloffay) [reply cmd/all-in-one/main.go]: I will try to move it in some package in collector. The same applies for tchannel there is duplication even in the current master.

File: cmd/all-in-one/Dockerfile (hunk 4 above)
[the remark this answers is not in the document]
Comment (yurishkuro) [reply cmd/all-in-one/Dockerfile]: I don't care which, as a long as it's unique.

We my want to move all ports into common package across all components so that it's easier to check for uniqueness.

File: model/proto/api_v2.proto [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    rpc PostSpans(PostSpansRequest) returns (PostSpansResponse) {
```
[the remark this answers is not in the document]
Comment (yurishkuro) [reply model/proto/api_v2.proto]: PS can be addressed in another PR, we should merge the infra first.

File: cmd/all-in-one/main.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (yurishkuro) [reply cmd/all-in-one/main.go]: But don't you think we'll need withxxx methods for grpc too? Eg we won't be able to use it internally unless we have discovery hook up WithDiscoverer

File: model/proto/api_v2.proto (hunk 7 above)
[the remark this answers is not in the document]
Comment (pavolloffay) [reply model/proto/api_v2.proto]: If it can I would prefer.

File: cmd/agent/app/reporter/grpc/flags.go [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+import (
+	"flag"
+
+	"github.com/spf13/viper"
+)
+
+const (
+	gRPCPrefix        = "reporter.grpc."
+	collectorHostPort = "collector.host-port"
```
[the remark this answers is not in the document]
Comment (pavolloffay) [reply cmd/agent/app/reporter/grpc/flags.go]: grpc dial supports only one address by default. An option is to use with balancer https://stackoverflow.com/a/50302295/4158442 not sure what that implies though.

**Comment by pavolloffay** [comment]
Two major outstanding tasks are:

- [ ] metrics

- [ ] rpc/stream for posting spans

I would prefer to do it in a separate PR if possible.

File: cmd/agent/app/reporter/grpc/reporter.go [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+func (r *Reporter) send(spans []*model.Span) error {
+	var process model.Process
+	if len(spans) > 0 {
+		process = *spans[0].Process
+	}
+	batch := model.Batch{Spans: spans, Process: process}
+	req := &api_v2.PostSpansRequest{Batch: batch}
+	_, err := r.collector.PostSpans(context.Background(), req)
```
[the remark this answers is not in the document]
Comment (pavolloffay) [reply cmd/agent/app/reporter/grpc/reporter.go]: Not sure at the moment. Now there is only one parameter `bool:ok` we might remove even that and add later what is needed

File: cmd/agent/app/reporter/grpc/reporter.go [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+func NewReporter(conn *grpc.ClientConn, logger *zap.Logger) *Reporter {
+	return &Reporter{
+		collector: api_v2.NewCollectorServiceClient(conn),
+		logger:    logger,
+	}
+}
+
+// EmitBatch implements EmitBatch() of Reporter
+func (r *Reporter) EmitBatch(b *thrift.Batch) error {
+	spans := jConverter.ToDomain(b.Spans, b.Process)
```
[the remark this answers is not in the document]
Comment (pavolloffay) [reply cmd/agent/app/reporter/grpc/reporter.go]: Had a quick look there is no to domain without process. I would prefer a separate PR this is already quite long.

[the remark this answers is not in the document]
Comment (pavolloffay) [reply cmd/agent/app/reporter/grpc/reporter.go]: I will add todo in the code and action in the PR comment

File: cmd/agent/app/reporter/grpc/flags.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (yurishkuro) [reply cmd/agent/app/reporter/grpc/flags.go]: the load balancer implementation can be done later, I'm just asking whether the cmd line should support a list of addresses (it can fail for now on >1 address).

**Comment by yurishkuro** [comment]
> I would prefer to do it in a separate PR if possible.

That's fine with me (I'd prefer that too, this diff is already too large).

But the main question we need to answer is what to do w.r.t. releasing 1.8. If there are any user-facing API changes (like supporting multiple addresses may change the cli flag name), then once we merge this PR we'll need to finish the outstanding tasks before cutting 1.8. Thoughts?

**Comment by pavolloffay** [comment]
Originally I wanted to cut 1.8 without gRPC to have more baking time in master before releasing it. I will have a look at the list of addresses.

**Comment by pavolloffay** [comment]
@yurishkuro I have added load balancer support in the last commit. If you prefer I can revert it, supporting a list of collectors in the same flag is back compatible anyways.

File: cmd/agent/app/reporter/grpc/collector_proxy.go [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+	// a way to fail immediately is to call WithBlock and WithTimeout
+	var conn *grpc.ClientConn
+	if len(o.CollectorHostPort) > 1 {
+		r, _ := manual.GenerateAndRegisterManualResolver()
+		var resolvedAddrs []resolver.Address
+		for _, addr := range o.CollectorHostPort {
+			resolvedAddrs = append(resolvedAddrs, resolver.Address{Addr: addr})
+		}
+		r.InitialAddrs(resolvedAddrs)
+		conn, _ = grpc.Dial(r.Scheme()+":///jaeger.collector", grpc.WithInsecure(), grpc.WithBalancerName(roundrobin.Name))
```
Comment (yurishkuro) [inline cmd/agent/app/reporter/grpc/collector_proxy.go]: what is the significance of `":///jaeger.collector"` string?

File: cmd/all-in-one/Dockerfile (hunk 4 above)
[the remark this answers is not in the document]
Comment (pavolloffay) [reply cmd/all-in-one/Dockerfile]: I have used 14250
